### DISCUSSION TITLE
Initializing a channel's pipeline in a `ChannelFactory`

### DISCUSSION THREAD

**Discussion by jon-signal** [netty/netty#13687] [Q&A] [answered]
Hello!

I'm trying to figure out if it's practical (or advisable) to try to perform non-trivial channel pipeline initialization inside a `ChannelFactory`. I am fully aware that this is not the first-choice path, but I'm trying to work around some unusual constraints.

As a greatly simplified example, let's say I'm working with a third-party library that uses a plaintext protocol, and we'd really strongly prefer to have it use TLS (or, as another example to avoid getting too focused on the specifics of TLS, a proxy). The library doesn't provide a direct way for us to customize channel pipelines, but it _does_ accept a custom `ChannelFactory`. I'm trying to figure out if we can (for example) create something like a `TlsChannelFactory` that yields channels with pipelines pre-configured to use TLS in a way that's transparent to the third-party library.

I've had some mixed results with early experiments, and I'm realizing that I'm wading into territory where somebody with more experience can just say more about the design intent and whether or not this is a terrible idea. So _is_ this a terrible idea? If it's doable, are there any pitfalls I should avoid or best practices I should adopt?

Thanks kindly!

**Comment by chrisvest** [comment, marked as answer]
The ChannelFactory is not a good place to initialise channels, because you'll be modifying the channel before it's registered with the event loop. If any part of the initialisation then wants to do delayed or scheduled work, it might break. You might be interested in something like https://github.com/netty/netty/pull/13565 as an alternative path forward. Those callbacks get a chance to modify the pipeline after its normal initialisation.

**Reply by jon-signal** [reply]
Thanks! That matches my understanding of channel initialization, and the `ChannelInitializerExtension` system from #13565 sounds like a very relevant (and timely!) option for us. I think we can call this The Answer™ for now.

As a follow-up question: do you have advice for how to apply a `ChannelInitializerExtension` selectively? If our application has multiple Netty-based libraries in play, how could we apply an extension to pipelines created by one third-party library, but not another? Reading the code, I think my current strategy would be to look for some tell-tale handler elsewhere in the pipeline (and I think that will work!), but wonder if there's something else you had in mind.

Cheers!

**Reply by chrisvest** [reply]
You can look for specific handlers, or you can mark the channel with an `Attribute`. The latter would not produce any false positives, but it might be difficult to arrange with 3rd party code.

**Reply by jon-signal** [reply]
…but in our case, we have the `ChannelFactory`, so I think that will work. Thank you very much—this is really helpful!
